metis-technical-metadata-generation/src/main/java/eu/europeana/metis/technical/metadata/generation/utilities/MediaExtractorForFile.java

```diff
@@ -0,0 +1,174 @@
+import eu.europeana.metis.mediaprocessing.model.RdfResourceEntry;
+import eu.europeana.metis.mediaprocessing.model.ResourceExtractionResult;
+import eu.europeana.metis.mediaprocessing.model.Thumbnail;
+import eu.europeana.metis.mediaprocessing.model.UrlType;
+import eu.europeana.metis.technical.metadata.generation.model.FileStatus;
+import eu.europeana.metis.technical.metadata.generation.model.TechnicalMetadataWrapper;
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Scanner;
+import java.util.concurrent.Callable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link Callable} that will go through a initialized dataset file and parse each resource per
+ * line.
+ * <p>This implementation saves the status of the file in the database so that on a subsequent
+ * execution, depending on the configuration, it will skip an X amount of lines. It can re-process
+ * failed resources or reset the reading of the file from the beginning, if set in the
+ * configuration.</p>
+ *
+ * @author Simon Tzanakis (Simon.Tzanakis@europeana.eu)
+ * @since 2019-04-17
+ */
+public class MediaExtractorForFile implements Callable<Void> {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MediaExtractorForFile.class);
+  private final File datasetFile;
+  private MongoDao mongoDao;
+  private MediaExtractor mediaExtractor;
+  private boolean startFromBeginningOfFiles;
+  private boolean retryFailedResources;
+
+  MediaExtractorForFile(File datasetFile, MongoDao mongoDao, MediaExtractor mediaExtractor,
+      boolean startFromBeginningOfFiles, boolean retryFailedResources) {
+    this.datasetFile = datasetFile;
+    this.mongoDao = mongoDao;
+    this.mediaExtractor = mediaExtractor;
+    this.startFromBeginningOfFiles = startFromBeginningOfFiles;
+    this.retryFailedResources = retryFailedResources;
+  }
+
+  @Override
+  public Void call() throws Exception {
+    return parseMediaForFile(datasetFile);
+  }
+
+  private Void parseMediaForFile(File datasetFile) throws IOException {
+    LOGGER.info(EXECUTION_LOGS_MARKER, "Parsing: {}", datasetFile);
+    final FileStatus fileStatus = getFileStatus(datasetFile.getName(), startFromBeginningOfFiles,
+        retryFailedResources);
+
+    int lineIndex = fileStatus.getLineReached();
+    try (Scanner scanner = new Scanner(datasetFile, "UTF-8")) {
+      //Bypass lines until the reached one, from a previous execution
+      if (moveScannerToLine(scanner, fileStatus)) {
+        while (scanner.hasNextLine()) {
+          String resourceUrl = scanner.nextLine();
+          LOGGER.info(EXECUTION_LOGS_MARKER, "Processing resource: {}", resourceUrl);
+          //Should this resource be processed according to status and configuration parameters
+          if (eligibleForProcessing(resourceUrl)) {
+            final ResourceExtractionResult resourceExtractionResult;
+            try {
```

Comment:
Change to:
try(final ResourceExtractionResult resourceExtractionResult = performMediaExtraction(resourceUrl)) {
  mongoDao.storeMediaResultInDb(resourceExtractionResult);
}
ResourceExtractionResult is a closable that closes all it's thumbnails inside, and this way the method is guaranteed to be called even if the mongoDao throws an exception or something.

Reply:
Sounds good